Ambiguous warning when main result is not Unit

Warning about def main(args: Array[String]) = ??? is ambiguous.

This is a trivial matter, yet it burned a minute to see whether the test that provokes it is actually run.  (I think all the runners use ScalaClassLoader.run.)

```scala
apm@mara:~/projects/snytt/test$ skalac -d foo files/run/repl-backticks.scala
files/run/repl-backticks.scala:3: warning: Test has a main method with parameter type Array[String], but Test will not be a runnable program.
  Reason: main method must have exact signature (Array[String])Unit
object Test {
       ^
one warning found
apm@mara:~/projects/snytt/test$ ls foo
Test.class  Test$.class
apm@mara:~/projects/snytt/test$ skala -cp foo Test
import java.lang.Thread.`yield`
import scala.`package`.Throwable
apm@mara:~/projects/snytt/test$ 
```

--- a/src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala
@@ -283,7 +283,7 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
             |""".stripMargin,
         WarningCategory.Other,
         sym)
-      val possibles = (sym.tpe nonPrivateMember nme.main).alternatives
+      val possibles      = sym.tpe.nonPrivateMember(nme.main).alternatives
       val hasApproximate = possibles.exists(m => cond(m.info) { case MethodType(p :: Nil, _) => p.tpe.typeSymbol == definitions.ArrayClass })
 
       // Before erasure so we can identify generic mains.
@@ -308,16 +308,17 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
         val mainAdvice =
           if (hasExact) Nil
           else possibles.map { m =>
-            m.info match {
+            val msg = m.info match {
               case PolyType(_, _) =>
-                ("main methods cannot be generic", m)
+                "main methods cannot be generic"
               case MethodType(params, res) if res.typeSymbol :: params exists (_.isAbstractType) =>
-                ("main methods cannot refer to type parameters or abstract types", m)
+                "main methods cannot refer to type parameters or abstract types"
               case MethodType(_, _) =>
-                ("main methods must have the exact signature (Array[String])Unit", m)
+                "main methods must have the exact signature (Array[String])Unit"
               case tp =>
-                (s"don't know what this is: $tp", m)
+                s"don't know what this is: $tp"
             }
+            (msg, m)
           }
 
         companionAdvice.foreach(msg => warnNoForwarder(msg, hasExact, exactly.fold(alternate)(_.info)))

--- a/src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala
+++ b/src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala
@@ -313,8 +313,10 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
                 "main methods cannot be generic"
               case MethodType(params, res) if res.typeSymbol :: params exists (_.isAbstractType) =>
                 "main methods cannot refer to type parameters or abstract types"
+              case MethodType(param :: Nil, _) if definitions.isArrayOfSymbol(param.tpe, StringClass) =>
+                "main methods must have the exact signature `(Array[String]): Unit`, though Scala runners will forgive a non-Unit result"
               case MethodType(_, _) =>
-                "main methods must have the exact signature (Array[String])Unit"
+                "main methods must have the exact signature `(Array[String]): Unit`"
               case tp =>
                 s"don't know what this is: $tp"
             }

--- a/src/reflect/scala/reflect/internal/Definitions.scala
+++ b/src/reflect/scala/reflect/internal/Definitions.scala
@@ -456,13 +456,13 @@ trait Definitions extends api.StandardDefinitions {
       else if (isScalaRepeatedParamType(tp)) elementExtract(RepeatedParamClass, tp) orElse tp
       else tp
     )
-    def repeatedToSingle(tp: Type): Type                     = elementExtract(RepeatedParamClass, tp) orElse elementExtract(JavaRepeatedParamClass, tp) orElse tp
+    def repeatedToSingle(tp: Type): Type               = elementExtract(RepeatedParamClass, tp) orElse elementExtract(JavaRepeatedParamClass, tp) orElse tp
      // We don't need to deal with JavaRepeatedParamClass here, as `repeatedToSeq` is only called in the patmat translation for Scala sources.
-    def repeatedToSeq(tp: Type): Type                        = elementTransform(RepeatedParamClass, tp)(seqType) orElse tp
-    def seqToRepeated(tp: Type): Type                        = elementTransform(SeqClass, tp)(scalaRepeatedType) orElse tp
-    def isReferenceArray(tp: Type)                           = elementTest(ArrayClass, tp)(elemtp => elemtp <:< AnyRefTpe || (elemtp eq ObjectTpeJava))
-    def isArrayOfSymbol(tp: Type, elem: Symbol)              = elementTest(ArrayClass, tp)(_.typeSymbol == elem)
-    def elementType(container: Symbol, tp: Type): Type       = elementExtract(container, tp)
+    def repeatedToSeq(tp: Type): Type                  = elementTransform(RepeatedParamClass, tp)(seqType) orElse tp
+    def seqToRepeated(tp: Type): Type                  = elementTransform(SeqClass, tp)(scalaRepeatedType) orElse tp
+    def isReferenceArray(tp: Type)                     = elementTest(ArrayClass, tp)(elemtp => elemtp <:< AnyRefTpe || (elemtp eq ObjectTpeJava))
+    def isArrayOfSymbol(tp: Type, elem: Symbol)        = elementTest(ArrayClass, tp)(_.typeSymbol == elem)
+    def elementType(container: Symbol, tp: Type): Type = elementExtract(container, tp)
 
     // Classes treated specially with respect to -Ywarn-unused
     lazy val SubTypeClass       = requiredClass[scala.<:<[_,_]]
@@ -474,7 +474,7 @@ trait Definitions extends api.StandardDefinitions {
     lazy val IteratorClass          = requiredClass[scala.collection.Iterator[_]]
     lazy val IterableClass          = requiredClass[scala.collection.Iterable[_]]
     lazy val ListClass              = requiredClass[scala.collection.immutable.List[_]]
-             def List_cons              = getMemberMethod(ListClass, nme.CONS)
+         def List_cons              = getMemberMethod(ListClass, nme.CONS)
     @migration("SeqClass now refers to scala.collection.immutable.Seq", "2.13.0")
     lazy val SeqClass               = requiredClass[scala.collection.immutable.Seq[_]]
     lazy val SeqFactoryClass        = requiredModule[scala.collection.SeqFactory.type]
@@ -640,8 +640,7 @@ trait Definitions extends api.StandardDefinitions {
       case _                            => false
     })
     // The given class has a main method.
-    def hasJavaMainMethod(sym: Symbol): Boolean =
-      (sym.tpe member nme.main).alternatives exists isJavaMainMethod
+    def hasJavaMainMethod(sym: Symbol): Boolean = sym.tpe.member(nme.main).alternatives.exists(isJavaMainMethod)
 
     class VarArityClass(name: String, maxArity: Int, countFrom: Int = 0, init: Option[ClassSymbol] = None) extends VarArityClassApi {
       private[this] val offset = countFrom - init.size

--- a/test/files/neg/main1.check
+++ b/test/files/neg/main1.check
@@ -29,7 +29,7 @@ main1.scala:41: warning: Foo has a valid main method (args: Array[String]): Unit
   object Foo extends Foo {  // Overrides main from the class
          ^
 main1.scala:53: warning: not a valid main method for p6.Main,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 
@@ -42,7 +42,7 @@ main1.scala:59: warning: Main has a main method (args: Array[Int]): Unit,
   object Main {
          ^
 main1.scala:60: warning: not a valid main method for p7.Main,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 
@@ -55,19 +55,26 @@ main1.scala:66: warning: Main has a main method,
   object Main {
          ^
 main1.scala:68: warning: not a valid main method for p8.Main,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 
     def main(args: Array[Double]) = ()
         ^
 main1.scala:67: warning: not a valid main method for p8.Main,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 
     def main(args: Array[Int]) = ()
         ^
+main1.scala:74: warning: not a valid main method for t7448.Main,
+  because main methods must have the exact signature `(Array[String]): Unit`, though Scala runners will forgive a non-Unit result.
+  To define an entry point, please define the main method as:
+    def main(args: Array[String]): Unit
+
+    def main(args: Array[String]) = ???
+        ^
 error: No warnings can be incurred under -Werror.
-11 warnings
+12 warnings
 1 error

--- a/test/files/neg/main1.scala
+++ b/test/files/neg/main1.scala
@@ -68,3 +68,9 @@ package p8 {
     def main(args: Array[Double]) = ()
   }
 }
+
+package t7448 {
+  object Main {
+    def main(args: Array[String]) = ???
+  }
+}

--- a/test/files/neg/t4749.check
+++ b/test/files/neg/t4749.check
@@ -1,5 +1,5 @@
 t4749.scala:5: warning: not a valid main method for bippy.Fail1,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`, though Scala runners will forgive a non-Unit result.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 
@@ -38,7 +38,7 @@ t4749.scala:28: warning: Fail6 has a valid main method (args: Array[String]): Un
   object Fail6 {
          ^
 t4749.scala:44: warning: not a valid main method for bippy.Win3,
-  because main methods must have the exact signature (Array[String])Unit.
+  because main methods must have the exact signature `(Array[String]): Unit`, though Scala runners will forgive a non-Unit result.
   To define an entry point, please define the main method as:
     def main(args: Array[String]): Unit
 

--- a/test/files/run/t7448.scala
+++ b/test/files/run/t7448.scala
@@ -1,0 +1,18 @@
+// scalac: -nowarn
+import util.chaining._
+
+object Test {
+  def main(args: Array[String]) = 42.tap(res => assert(res == 42))
+}
+
+// test that partest is using scala runner to execute this test.
+// With warnings enabled:
+/*
+t7448.scala:7: warning: not a valid main method for Test,
+  because main methods must have the exact signature `(Array[String]): Unit`, though Scala runners will forgive a non-Unit result.
+  To define an entry point, please define the main method as:
+    def main(args: Array[String]): Unit
+
+  def main(args: Array[String]) = 42.tap(res => assert(res == 42))
+      ^
+ */

